### DISCUSSION TITLE
FHIR nullFlavor

### DISCUSSION THREAD

**Discussion by maximalexeev** [fhir-ru/core#180] [Административная РГ]
Обсуждение nullFlavor 

Для начала - цитата @shamilnn  отсюда https://github.com/fhir-ru/core/discussions/38#discussioncomment-1852222

_Это может зависеть от типа данных элемента и cardinality.

Например, если элемент обязательный (1..1, 1..*), тип данных CodeableConcept и binding [Extensible](http://hl7.org/fhir/R4/terminologies.html#extensible) или выше (Required), то элемент ДОЛЖЕН быть заполнен в соответствии с предлагаемым ValueSet. Если, при отсутствии информации можно передавать NullFlavor, то необходимо включить код(ы) NullFlavor в ValueSet подобного элемента.

Если информации нет, и элемент опциональный (0..1, 0..*), то FHIR запрещает передавать пустой элемент (ситуация, когда при отсутствии информации нельзя передавать элемент)._

**Comment by Vrodvi** [comment]
Уже обсуждали и даже принято вполне разумное решение:
http://fhir-ru.zendoc.me/meetings.2022-04-26
> 47:27 использование NullFlavor – решили следовать рекомендациям FHIR: если есть атрибут для указания причины отсутствия значения (например, в Observation) то используем его, а если нет, и тип атрибута CodeableConcept, то используем или добавляем в него значения, отражающие причины отсутствия информации; для иных типов данных используем стандартное расширение NullFlavor с жесткой привязкой к значениям v3.

Думаю стоит уточнить описание дискуссии или дать комментарий с конкретными вопросами для обсуждения.

**Comment by JustAsWalker** [comment]
По результатам встречи http://fhir-ru.zendoc.me/meetings.2022-11-01 к решению от 26/04 возникло несколько вопросов.

Во-первых, [Observation.dataAbsentReason](https://www.hl7.org/fhir/observation-definitions.html#Observation.dataAbsentReason) описывает отсутствие именно элемента value, но если неким профилем будет затребовано обязательное наличие другого атрибута (например, материала), то как передать NullFlavor в данном случае? По всей видимости расширением, но тогда получается не совсем последовательное решение - без дополнительной аналитики не понять, где искать отметку об отсутствии данных. Есть предложение не использовать для NullFlavor отдельные атрибуты, а всегда передавать отметку в виде расширения.

Во-вторых, в решении указано "_с жесткой привязкой к значениям v3_", но, если я правильно понял коллег, есть потребность передавать причину отсутствия информации по [справочнику ФНСИ](https://nsi.rosminzdrav.ru/#!/refbook/1.2.643.5.1.13.13.99.2.286/version/3.1). Так как [стандартное расширение](https://www.hl7.org/fhir/extension-iso21090-nullflavor.html) имеет результат с типом данных code (даже не CodeableConcept), в него мы данные из справочника ФНСИ поместить не можем. Обсуждались два варианта:
1. Создать собственное ru-core расширение по аналогии со стандартным и в данных писать оба
``` json
{
  "extension": [
    {
      "url": "http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor",
      "valueCode": "NA"
    },
    {
      "url": "http://fhir.ru/core/sd/core-ex-nullFlavor",
      "valueCoding": {
        "system": "urn:oid:1.2.643.5.1.13.13.99.2.286",
        "code": "NA",
        "display": "Неприменимо",
        "version": "3.1"
      }
    }
  ]
}
``` 

2. Создать собственное ru-core расширение и им дополнять значение из стандартного расширения (по аналогии с тем, как мы это сделали для [Patient.gender](http://fhir-ru.zendoc.me/core.Patient#:elements:gender)) - _хоть я и являюсь автором данного предложения, пока не совсем уверен, что валидация fhir позволит так делать_
``` json
{
  "extension": [
    {
      "url": "http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor",
      "valueCode": "NA",
      "_valueCode": {
        "extension": [
          {
            "url": "http://fhir.ru/core/sd/core-ex-nullFlavor",
            "valueCoding": {
              "system": "urn:oid:1.2.643.5.1.13.13.99.2.286",
              "code": "NA",
              "display": "Неприменимо",
              "version": "3.1"
            }
          }
        ]
      }
    }
  ]
}
```
